<?xml version="1.0"?>
<robot name="robot_arm">

    <link name="world"></link>

    <joint name="base_joint" type="fixed">
        <origin xyz="1.5 1.0 0" rpy="0 0 0"/>
        <parent link="world"/>
        <child link="base"/>
    </joint>

    <link name="base">
        <visual>
            <geometry>
                <box size="2.0 1.0 0.5"/>
            </geometry>
            <material name="light_blue">
                <color rgba="0.5 0.5 1.0 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="2.0 1.0 0.5"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0" izz="1"/>
        </inertial>
    </link>

    <link name="lower_body">
        <visual>
            <geometry>
                <cylinder radius="0.25" length="1.0"/>
            </geometry>
            <material name="dark_green">
                <color rgba="0.0 0.5 0.0 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.25" length="1.0"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <!-- Inertia tensor for a cylinder with mass=1kg, radius=0.25, and length=0.5 -->
            <inertia ixx="0.034375" ixy="0" ixz="0" iyy="0.034375" iyz="0" izz="0.0171875"/>
        </inertial>
    </link>

    <joint name="servo0" type="revolute">
        <parent link="base"/>
        <child link="lower_body"/>
        <origin xyz="0.5 0 0.5" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1.57" upper="1.57" effort="10000" velocity="1.0"/>
    </joint>

    <link name="middle_body">
        <visual>
            <origin xyz="0 0 0.5" rpy="0 0 0"/>
            <geometry>
                <box size="0.3 0.3 1.4"/>
            </geometry>
            <material name="grey">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0.5" rpy="0 0 0"/>
            <geometry>
                <box size="0.3 0.3 1.4"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
        </inertial>
    </link>

    <joint name="servo1" type="revolute">
        <parent link="lower_body"/>
        <child link="middle_body"/>
        <origin xyz="0 0 0.5" rpy="0 0 0"/>
        <axis xyz="0 -1 0"/>
        <limit lower="0.0" upper="1.57" effort="10000" velocity="1.0"/>
    </joint>

    <gazebo>
        <plugin filename="/opt/ros/humble/lib/libign_ros2_control-system.so" name="ign_ros2_control::IgnitionROS2ControlPlugin">
            <robot_param>robot_description</robot_param>
            <robot_param_node>robot_state_publisher</robot_param_node>
            <parameters>/home/small-thinking/mnlm/mnlm/robot/robot_arm_ws/install/robot_arm/share/robot_arm/config/robot_arm_control.yaml</parameters>
        </plugin>
    </gazebo>

    <ros2_control name="GazeboSimSystem" type="system">
        <hardware>
            <plugin>ign_ros2_control/IgnitionSystem</plugin>
        </hardware>
        <joint name="servo0">
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!-- <state_interface name="effort"/> -->
        </joint>
        <joint name="servo1">
            <command_interface name="position"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <!-- <state_interface name="effort"/> -->
        </joint>
    </ros2_control>

</robot>